app/src/main/java/org/dhis2/usescases/teiDashboard/dashboardfragments/teidata/TeiDataRepositoryImpl.kt

```diff
@@ -218,8 +221,26 @@ class TeiDataRepositoryImpl(
         }
     }
 
-    private fun checkAddEvent(): Boolean {
+    private fun checkAddEvent(stage: ProgramStage, isSelected: Boolean): Boolean {
         val enrollment = d2.enrollmentModule().enrollments().uid(enrollmentUid).blockingGet()
-        return !(enrollment == null || enrollment.status() != EnrollmentStatus.ACTIVE)
+        val enrollmentStatusCheck =
+            !(enrollment == null || enrollment.status() != EnrollmentStatus.ACTIVE)
+        val totalEventCount = d2.eventModule().events()
+            .byEnrollmentUid().eq(enrollmentUid)
+            .byProgramStageUid().eq(stage.uid())
+            .byDeleted().isFalse
+            .blockingCount()
+        val stageNotRepeatableZeroCount = stage.repeatable() != true &&
+            totalEventCount == 0
+        val stageRepeatableZeroCount = stage.repeatable() == true &&
+            totalEventCount == 0
+        val stageRepeatableCountSelected = stage.repeatable() == true &&
+            totalEventCount > 0 && isSelected
+
+        return enrollmentStatusCheck && (
+            stageNotRepeatableZeroCount ||
+                stageRepeatableZeroCount ||
+                stageRepeatableCountSelected
+            )
```

Comment:
Can you extract this to a class for moving to SDK :dancing_women: ? Thanks

Reply:
done